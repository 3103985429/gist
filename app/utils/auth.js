const jwt = require('jsonwebtoken');
const winston = require('winston');

const Users = require('../models/users');
const Tokens = require('../models/tokens');
const { extractHostname } = require('../../shared/url');
const { UserError } = require('../libs/system/errors');
const { handleError } = require('../utils/requests');

let ldapConfig;

if (process.env.AUTH_PROTOCOL === 'LDAP') {
    try {
        const { ldapConfig: customLdapConfig } = require(`../../user/${process.env.SUBDOMAIN}/src/auth`);
        ldapConfig = Object.assign({}, getDefaultLdapConfig(), customLdapConfig);
    } catch (e) {
        ldapConfig = getDefaultLdapConfig();
    }
}


function getDefaultLdapConfig() {
    return {
        getProfileDetails(profile) {
            let { firstName = '', lastName = '' } = profile;
            const email = profile.email || '';

            if (!firstName && profile.name) {
                [firstName, lastName] = profile.name.split(' ');
            }

            return { firstName, lastName, email };
        },
        getStrategyConfig() {
            return {};
        },
        onAfterLogin(req, res, profile, callback) {
            callback();
        },
    };
}

const appendTokenToRequest = (req, res, next) => {
    try {
        const { token } = req.query;
        req.token = token;

        next();
    } catch (error) {
        winston.error(error);

        return res.status(404).render('error', {});
    }
};

const loginWithToken = async (req, res, next) => {
    try {
        const token = await Tokens.findOne({ token: req.token || req.params.token }).populate('apiKey');

        if (!token) {
            req.flash('error', { message: 'Token is invalid.' });

            return res.redirect('/auth/login');
        }

        const referer = req.header('Referer');

        if (!referer) {
            return handleError(new UserError('Missing referer'), res);
        }

        const isAuthorisedDomain = token.apiKey.requestDomains.some(domain => extractHostname(referer) === domain);

        if (!isAuthorisedDomain) {
            return handleError(new UserError('This domain is not authorised.'), res);
        }

        const decoded = jwt.verify(token.token, process.env.SESSION_SECRET);
        const user = await Users.findById(decoded.userId);

        if (!user.defaultLoginTeam || !user._team.length || token.apiKey.team.toString() !== user.defaultLoginTeam.toString()) {
            req.flash('error', { message: 'User is not associated with a team.' });
            return res.redirect('/auth/login');
        }

        req.logIn(user, async (err) => {
            if (err) {
                return next(err);
            }

            token.usedAt = new Date();
            await token.save();

            req.user = user;

            if (req.token) {
                return next();
            } else {
                return res.redirect('/dashboard');
            }

        });
    } catch (e) {
        req.flash('error', { message: 'Token is invalid.' });
        return res.redirect('/auth/login');
    }
};

module.exports = {
    ldapConfig,
    appendTokenToRequest,
    loginWithToken,
};
