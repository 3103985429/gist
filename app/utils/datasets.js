const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
