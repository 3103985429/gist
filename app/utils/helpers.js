const { isEqual, trim, toString } = require('lodash');

module.exports.equals = a => b => isEqual(a, b);

module.exports.isEmbed = (query) => query.embed === 'true' || query.embed === true;

module.exports.formatColumnName = (value) => trim(toString(value).replace(/\./g, '_'));

module.exports.formatRawValues = (value) => trim(toString(value).replace(/’/g, '\'').replace(/”/g, '"'));

module.exports.addSlugIndex = (slugs, item, tilteProp = 'title') => {
    if (!slugs.length) {
        return;
    }

    let i = 2;
    const slugList = slugs.map((item) => item.slug);
    const max = i + slugList.length + 1;

    for (i; i < max; i++) {
        const newSlug = `${item.slug}-${i}`;

        if (!slugList.includes(newSlug)) {
            item.slug = newSlug;
            item[tilteProp] = `${item[tilteProp]} ${i}`;

            break;
        }
    }
};
