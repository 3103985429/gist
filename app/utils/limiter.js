const RateLimit = require('express-rate-limit');
const MongoStore = require('rate-limit-mongo');
const { limiterStore } = require('./mongoStore');

const dbURI = process.env.MONGODB_URI;

const windowTime = 15 * 60 * 1000; //15 minutes

const limiter = new RateLimit({
    store: limiterStore,
    windowMs: windowTime,
    max: 100, // 100 requests per 15 minutes
});

module.exports = limiter;
