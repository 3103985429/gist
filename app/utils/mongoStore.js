const MongoStore = require('rate-limit-mongo');
const winston = require('winston');

const dbURI = process.env.MONGODB_URI;

const limiterStore = new MongoStore({
    uri: dbURI,
    collectionName: 'expressRateRecords',
    errorHandler: winston.error,
});


const slowDownStore = new MongoStore({
    uri: dbURI,
    collectionName: 'slowDown',
    errorHandler: winston.error,
});


module.exports = {
    limiterStore,
    slowDownStore,
};
