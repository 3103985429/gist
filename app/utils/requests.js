const winston = require('winston');

const { PublicError } = require('../libs/system/errors');

module.exports = {
    handleError,
};

/**
 * Returns true when an error occurred.
 * @param {Error} err
 * @param {Object} res
 * @param {String=} message
 * @returns {Boolean}
 */
function handleError(err, res, message = 'An unexpected error happened during request.') {
    if (err) {
        const code = err.statusCode || 500;
        let responseMessage = message;

        if (err instanceof PublicError) {
            responseMessage = err.message;
        } else if (['development', 'testing'].includes(process.env.NODE_ENV)) {
            responseMessage = `Non-public error: ${err.message}`;
        }

        winston.error(message, err);
        res.status(code).json({ error: responseMessage });

        return true;
    }

    return false;
}
