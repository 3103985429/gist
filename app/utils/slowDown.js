const winston = require('winston');
const SlowDown = require('express-slow-down');
const { slowDownStore } = require('./mongoStore');

const windowTime = 15 * 60 * 1000; //15 minutes

const slowDown = (delayAfter = 20, windowTimeMs = windowTime) => new SlowDown({
    store: slowDownStore,
    windowMs: windowTimeMs,
    delayAfter,
    delayMs: 500,
    onLimitReached: (req) => {
        winston.error(`Throttled path: ${req.route.path}`);
        winston.error(`Throttled ip: ${req.headers['x-forwarded-for'] || req.connection.remoteAddress}`);
    }
});

module.exports = slowDown;
