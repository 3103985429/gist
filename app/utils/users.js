const Users = require('../models/users');

module.exports = {
    getUser,
};

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getUser(id) {
    return await Users
        .findById(id)
        .lean()
        .exec();
}
