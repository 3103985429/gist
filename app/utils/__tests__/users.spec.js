import * as users from '../users';
import { hydrateDb } from '../../../internals/testing/backend/utils';
import { user1 } from '../../../internals/testing/backend/fixtures/users';


describe('Users Utils', () => {
    beforeEach(async () => {
        await hydrateDb();
    });

    describe('when #getUser is called', () => {
        it('should return a user', async () => {
            const result = await users.getUser(user1._id);

            expect(result).toMatchSnapshot();
        });
    });
});
