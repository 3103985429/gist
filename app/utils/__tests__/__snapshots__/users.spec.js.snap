// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Users Utils when #getUser is called should return a user 1`] = `
Object {
  "__v": 0,
  "_articleEditors": Array [],
  "_articleViewers": Array [],
  "_editors": Array [],
  "_id": "5a42bf07d3bc0b20648290e0",
  "_siteEditors": Array [],
  "_siteViewers": Array [],
  "_team": Array [
    "5a42bf30d3bc0b20648290e1",
  ],
  "_viewers": Array [],
  "activated": true,
  "active": true,
  "canCreateNewArticle": Array [],
  "canCreateNewSite": Array [],
  "canCreateNewViz": Array [],
  "createdAt": 2017-12-26T21:28:39.946Z,
  "defaultLoginTeam": "5a42bf30d3bc0b20648290e1",
  "email": "user@schemadesign.com",
  "firstName": "Maitland",
  "hash": "dc449e96c930ca46653ec867c448719fab39dd85c4b91c1643d2faad140f24e711da13133e34ee4da0d3a5a2ca154edd4bfaeb9cf80cb9af57bce4d7ea1b94ce",
  "lastName": "McConnell",
  "lockUntil": 1,
  "loginAttempts": 0,
  "provider": "local",
  "salt": "test",
  "sampleImported": false,
  "signedUp": false,
  "updatedAt": 2017-12-26T21:36:31.420Z,
}
`;
